File: kirun-js/src/engine/util/string/StringBuilder.ts
import { KIRunConstants } from '../../constant/KIRunConstants';
import { KIRuntimeException } from '../../exception/KIRuntimeException';

export class StringBuilder {
    private str: string;

    public constructor(str?: string) {
        this.str = str ?? '';
    }

    public append(x: any): StringBuilder {
        this.str += x;
        return this;
    }

    public toString(): string {
        return '' + this.str;
    }

    public trim(): StringBuilder {
        this.str = this.str.trim();
        return this;
    }

    public setLength(num: number): StringBuilder {
        this.str = this.str.substring(0, num);
        return this;
    }

    public length(): number {
        return this.str.length;
    }

    public charAt(index: number): string {
        return this.str.charAt(index);
    }

    public deleteCharAt(index: number): StringBuilder {
        this.checkIndex(index);
        this.str = this.str.substring(0, index) + this.str.substring(index + 1);
        return this;
    }

    public insert(index: number, str: string): StringBuilder {
        this.str = this.str.substring(0, index) + str + this.str.substring(index);
        return this;
    }

    private checkIndex(index: number): void {
        if (index >= this.str.length)
            throw new KIRuntimeException(
                `Index ${index} is greater than or equal to ${this.str.length}`,
            );
    }

    public substring(start: number, end: number): string {
        return this.str.substring(start, end);
    }
}

File: kirun-py/src/kirun_py/util/string/string_builder.py
from __future__ import annotations
from typing import Any

from kirun_py.exception import KIRuntimeException


class StringBuilder:
    def __init__(self, s: str = ''):
        self._str = s

    def append(self, x: Any) -> StringBuilder:
        self._str += str(x)
        return self

    def __str__(self) -> str:
        return self._str

    def trim(self) -> StringBuilder:
        self._str = self._str.strip()
        return self

    def set_length(self, num: int) -> StringBuilder:
        self._str = self._str[:num]
        return self

    def length(self) -> int:
        return len(self._str)

    def char_at(self, index: int) -> str:
        return self._str[index]

    def delete_char_at(self, index: int) -> StringBuilder:
        self._check_index(index)
        self._str = self._str[:index] + self._str[index + 1:]
        return self

    def insert(self, index: int, s: str) -> StringBuilder:
        self._str = self._str[:index] + s + self._str[index:]
        return self

    def _check_index(self, index: int) -> None:
        if index >= len(self._str):
            raise KIRuntimeException(
                f'Index {index} is greater than or equal to {len(self._str)}'
            )

    def substring(self, start: int, end: int) -> str:
        return self._str[start:end]
